---
layout: post
published: false
draft: true
tags: R, cookbook, rtracklayer, GenomicRanges 
status: process
title: "How to import narrowPeak files"
author: "Charles Joly Beauparlant"
output: html_document
---

The goal of this post is to show how to import [narrowPeak](https://genome.ucsc.edu/FAQ/FAQformat.html#format12) and [broadPeak](https://genome.ucsc.edu/FAQ/FAQformat.html#format13) files into R in a valid `GRanges` format.

To do so, you need to have installed the [`rtracklayer`](http://bioconductor.org/packages/release/bioc/html/rtracklayer.html) package. To replicate the examples in this document, you will also need to install the [`GenomicFormatExamples`](https://github.com/CharlesJB/GenomicFormatExamples) package.

```{r library, message=FALSE}
require(rtracklayer)
require(GenomicFormatExamples)
```

# TL;DR

```{r tldr, eval=FALSE}
# To import narrowPeak files
extraCols_narrowPeak <- c(signalValue = "numeric", pValue = "numeric",
                          qValue = "numeric", peak = "integer")
gr_narrowPeak <- import(file_narrowPeak, format = "BED",
                        extraCols = extraCols_narrowPeak)

# To import broadPeak files
extraCols_broadPeak <- c(signalValue = "numeric", pValue = "numeric",
                          qValue = "numeric")
gr_broadPeak <- import.bed(file_broadPeak, format = "BED",
                           extraCols = extraCols_narrowPeak)
```

# The `rtracklayer` package

The `rtracklayer` package offers multiple ways to easily import various genomic formats such as BED, WIG or GFF/GTF. For instance, if we want to import a BED file we can use the `import` function:

```{r importBed}
bed_file <- get_demo_file(format = "bed")
gr_bed <- import(bed_file)
gr_bed
```

The `import` function can be used to import the following file formats:

* GFF
* BED
* Bed15
* bedGraph
* WIG
* BigWig

As shown in the previous example, the file format is derived from the file extension which is why it generally works correctly without have to specify the format.

# Importing narrowPeak and broadPeak

Unfortunately, the [narrowPeak](https://genome.ucsc.edu/FAQ/FAQformat.html#format12) and [broadPeak](https://genome.ucsc.edu/FAQ/FAQformat.html#format13) are not directly supported by the `import` function:

```{r narrowPeakBroadPeakFail}
narrowPeak_file <- get_demo_file(format = "narrowPeak")
import(narrowPeak_file)
broadPeak_file <- get_demo_file(format = "broadPeak")
import(broadPeak_file)
```

Even if we specify the format to be BED, the `import` function will fail:

```{r specifyFormatFail}
import(narrowPeak_file, format = "BED")
import(broadPeak_file, format = "BED")
```

The reason is that the `import` function checks the format of the content of every columns to make sure the file is in the good format and that the columns in BED files are not completely identical to those in narrowPeak/broadPeak files.

BED files:

1. chrom 
2. chromStart
3. chromEnd
4. name
5. score
6. strand
7. thickStart
8. thickEnd
9. itemRgb
10. blockCount
11. blockSizes
12. blockStarts

narrowPeak/broadPeak files:

1. chrom
2. chromStart
3. chromEnd
4. name
5. score
6. strand
7. signalValue
8. pValue
9. qValue
10. peak (narrowPeak only)

The first 6 columns are the same, but the seventh column is different. In the BED format, the `thickStart` is an integer while in the narrowPeak/broadPeak format the `signalValue` is a numeric.

In order to solve this problem, we need to use the `extraCols` parameter:

```
extraCols: A character vector in the same form as ‘colClasses’ from
	‘read.table’.  It should indicate the name and class of each
	extra/special column to read from the BED file. As BED does
	not encode column names, these are assumed to be the last
	columns in the file. This enables parsing of the various
	BEDX+Y formats.
```

In other words, we need to give the name and type of every columns starting at the one that is different from the standard BED format. In our case, we need to give the name and type of the 7th, 8th, 9th and 10th (in the case of narrowPeak) columns. To to so, we have to create a named vector:

```{r extraColsNarrow}
extraCols_narrowPeak <- c(singnalValue = "numeric", pValue = "numeric",
                          qValue = "numeric", peak = "integer")
gr_narrowPeak <- import(narrowPeak_file, format = "BED",
                        extraCols = extraCols_narrowPeak)
gr_narrowPeak
```

```{r extraColsBroad}
extraCols_broadPeak <- c(singnalValue = "numeric", pValue = "numeric",
                         qValue = "numeric", peak = "integer")
gr_broadPeak <- import(broadPeak_file, format = "BED",
                       extraCols = extraCols_broadPeak)
gr_broadPeak
```

# Conclusion

It is very important to avoid to re-invent the wheel. Since the narrowPeak and the broadPeak are not directly supported by the `import` function in the `rtracklayer` package, we could be tempted to import them manually with the `read.table` function.

Not only would this be more complicated because we need to add the names of all the columns before converting in the `GRanges` format, but more importantly we need to make sure to convert the 0-based coordinate system of the BED file to the 1-based coordinate system of the `GRanges`:

```{r readTableBed}
tbl_bed <- read.table(bed_file, header = FALSE)
head(tbl_bed)
gr_bed
```

Notice how the start value is different depending of the strategy used to import the file. The correct one is the one obtained with the `import` function.
